### PR TITLE
 Support shared refcount for Objective-C references to Kotlin objects

### PR DESCRIPTION
**Pull request by SvyatoslavScherbina**
Also add missing test for Objective-C weak refs to Kotlin objects

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/BlockPointerSupport.kt [hunk 1]
```diff
@@ -136,9 +136,11 @@ private val BlockPointerBridge.nameSuffix: String
 internal class BlockAdapterToFunctionGenerator(val objCExportCodeGenerator: ObjCExportCodeGenerator) {
     private val codegen get() = objCExportCodeGenerator.codegen
 
+    private val kRefSharedHolderType = LLVMGetTypeByName(codegen.runtime.llvmModule, "class.KRefSharedHolder")!!
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/BlockPointerSupport.kt]: Hmm, indeed `class.`?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/BlockPointerSupport.kt [hunk 2]
```diff
@@ -163,15 +165,22 @@ internal class BlockAdapterToFunctionGenerator(val objCExportCodeGenerator: ObjC
...
 
         val srcBlockPtr = bitcast(pointerType(blockLiteralType), param(1))
-        val srcSlot = structGep(srcBlockPtr, 1)
+        val srcRefHolder = structGep(srcBlockPtr, 1)
+
+        // Note: in current implementation copy helper is invoked only for stack-allocated blocks from the same thread,
+        // so it is technically not necessary to check owner.
+        // However this is not guaranteed by Objective-C runtime, so keep it suboptimal but reliable:
+        val ref = call(
+                context.llvm.kRefSharedHolderRef,
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/BlockPointerSupport.kt]: Is it inlined on machine code level?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/BlockPointerSupport.kt (hunk 1 above)
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/BlockPointerSupport.kt]: That's the name of the LLVM type produced by clang for `class KRefSharedHolder`.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/BlockPointerSupport.kt (hunk 2 above)
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/BlockPointerSupport.kt]: No. Should it be?

**Pull request merged**